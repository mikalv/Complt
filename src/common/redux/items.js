import {
  DELETE_ITEM_POUCH,
  INSERT_ITEM_POUCH,
  UPDATE_ITEM_POUCH,
  CREATE_ITEM,
  COMPLETE_ITEM,
  DELETE_ITEM,
  DELETE_ITEM_WITHOUT_PARENT,
  UPDATE_ITEM,
  MOVE_ITEM,
  MOVE_ITEM_WITHOUT_PARENT,
  REORDER_ITEM,
} from './actionTypes';

function arrayMove(arr, previousIndex, newIndex) {
  const array = arr.slice(0);
  let safeNewIndex = newIndex;
  if (newIndex >= array.length) {
    safeNewIndex = array.length - 1;
  }
  array.splice(safeNewIndex, 0, array.splice(previousIndex, 1)[0]);
  return array;
}
export const initialState = [];

function ensureRootAndInboxExists(state, action) {
  if (action.parentProjectId === 'root') {
    if (state.findIndex(item => item._id === 'root') !== -1) return state;
    return [...state, { _id: 'root', isProject: true, children: [] }];
  }
  if (action.parentProjectId === 'inbox') {
    if (state.findIndex(item => item._id === 'inbox') !== -1) return state;
    return [...state, { _id: 'inbox', isProject: true, children: [] }];
  }
  return state;
}

export default function itemsReducer(state = initialState, action) {
  switch (action.type) {
    case DELETE_ITEM_WITHOUT_PARENT:
    case DELETE_ITEM_POUCH: {
      const indexOfId = state.findIndex(item => item._id === action.id);
      if (indexOfId === -1) return state;
      return [...state.slice(0, indexOfId), ...state.slice(indexOfId + 1)];
    }
    case INSERT_ITEM_POUCH: {
      return [...state, action.item];
    }
    case UPDATE_ITEM_POUCH: {
      const indexOfId = state.findIndex(item => item._id === action.item._id);
      if (indexOfId === -1) return state;
      return [
        ...state.slice(0, indexOfId),
        action.item,
        ...state.slice(indexOfId + 1),
      ];
    }
    case CREATE_ITEM: {
      const newState = ensureRootAndInboxExists(state, action);
      const parentProjectIndex = newState.findIndex(
        item => item._id === action.parentProjectId
      );
      return [
        ...newState.slice(0, parentProjectIndex),
        {
          ...newState[parentProjectIndex],
          children: [...newState[parentProjectIndex].children, action.item._id],
        },
        ...newState.slice(parentProjectIndex + 1),
        action.item,
      ];
    }
    case COMPLETE_ITEM: {
      const indexOfId = state.findIndex(item => item._id === action.id);
      if (indexOfId === -1) return state;
      return [
        ...state.slice(0, indexOfId),
        {
          ...state[indexOfId],
          isCompleted: action.isCompleted,
        },
        ...state.slice(indexOfId + 1),
      ];
    }
    case DELETE_ITEM: {
      const parentProject = state.find(
        item => item._id === action.parentProjectId
      );
      const childIndexInParent = parentProject.children.indexOf(action.id);
      if (childIndexInParent === -1) return state;
      const childIndex = state.findIndex(item => item._id === action.id);
      if (state[childIndex].isProject && state[childIndex].children.length > 0)
        return state;
      const stateWithoutTask = [
        ...state.slice(0, childIndex),
        ...state.slice(childIndex + 1),
      ];
      const parentProjectIndex = stateWithoutTask.findIndex(
        item => item._id === action.parentProjectId
      );
      return [
        ...stateWithoutTask.slice(0, parentProjectIndex),
        {
          ...parentProject,
          children: [
            ...parentProject.children.slice(0, childIndexInParent),
            ...parentProject.children.slice(childIndexInParent + 1),
          ],
        },
        ...stateWithoutTask.slice(parentProjectIndex + 1),
      ];
    }
    case UPDATE_ITEM: {
      const indexOfOldItem = state.findIndex(
        item => item._id === action.item._id
      );
      const oldItem = state[indexOfOldItem];
      return [
        ...state.slice(0, indexOfOldItem),
        {
          ...oldItem,
          name: action.item.name,
          tags: action.item.tags,
          dates: action.item.dates,
        },
        ...state.slice(indexOfOldItem + 1),
      ];
    }
    case MOVE_ITEM: {
      if (action.previousParent === action.newParent) return state;
      const itemIndex = state.findIndex(item => item._id === action.id);
      const previousParentIndex = state.findIndex(
        item => item._id === action.previousParent
      );
      const newParentIndex = state.findIndex(
        item => item._id === action.newParent
      );
      if (
        itemIndex === -1 ||
        previousParentIndex === -1 ||
        newParentIndex === -1
      )
        return state;
      if (
        !state[newParentIndex].isProject ||
        !state[previousParentIndex].isProject
      )
        return state;
      const previousParent = state[previousParentIndex];
      const childIndexInPreviousParent = previousParent.children.indexOf(
        action.id
      );
      if (childIndexInPreviousParent === -1) return state;
      const stateWithoutItemInPreviousParent = [
        ...state.slice(0, previousParentIndex),
        {
          ...previousParent,
          children: [
            ...previousParent.children.slice(0, childIndexInPreviousParent),
            ...previousParent.children.slice(childIndexInPreviousParent + 1),
          ],
        },
        ...state.slice(previousParentIndex + 1),
      ];
      const newParent = state[newParentIndex];
      return [
        ...stateWithoutItemInPreviousParent.slice(0, newParentIndex),
        { ...newParent, children: [...newParent.children, action.id] },
        ...stateWithoutItemInPreviousParent.slice(newParentIndex + 1),
      ];
    }
    case MOVE_ITEM_WITHOUT_PARENT: {
      const newParentIndex = state.findIndex(
        item => item._id === action.newParent
      );
      const itemIndex = state.findIndex(item => item._id === action.id);
      if (itemIndex === -1 || newParentIndex === -1) return state;
      const newParent = state[newParentIndex];
      if (!newParent.isProject) return state;
      return [
        ...state.slice(0, newParentIndex),
        { ...newParent, children: [...newParent.children, action.id] },
        ...state.slice(newParentIndex + 1),
      ];
    }
    case REORDER_ITEM: {
      if (action.oldIndex === action.newIndex) return state;
      const projectIndex = state.findIndex(item => item._id === action.id);
      const project = state[projectIndex];
      const newProjectChildren = arrayMove(
        project.children,
        action.oldIndex,
        action.newIndex
      );
      return [
        ...state.slice(0, projectIndex),
        { ...project, children: newProjectChildren },
        ...state.slice(projectIndex + 1),
      ];
    }
    default: {
      return state;
    }
  }
}
