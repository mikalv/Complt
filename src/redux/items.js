import {
  DELETE_ITEM_POUCH,
  INSERT_ITEM_POUCH,
  UPDATE_ITEM_POUCH,
  CREATE_ITEM,
  COMPLETE_TASK,
  DELETE_TASK,
  DELETE_PROJECT,
  UPDATE_ITEM,
  MOVE_ITEM,
} from './actionTypes';

export const initialState = [];

function ensureRootAndInboxExists(state, action) {
  if (action.parentProjectId === 'root') {
    if (state.findIndex(item => item._id === 'root') !== -1) return state;
    return [
      ...state,
      { _id: 'root', isProject: true, children: [] },
    ];
  }
  if (action.parentProjectId === 'inbox') {
    if (state.findIndex(item => item._id === 'inbox') !== -1) return state;
    return [
      ...state,
      { _id: 'inbox', isProject: true, children: [] },
    ];
  }
  return state;
}

export default function itemsReducer(state = initialState, action) {
  switch (action.type) {
    case DELETE_ITEM_POUCH: {
      const indexOfId = state.findIndex(item => item._id === action.id);
      if (indexOfId === -1) return state;
      return [
        ...state.slice(0, indexOfId),
        ...state.slice(indexOfId + 1),
      ];
    }
    case INSERT_ITEM_POUCH: {
      return [
        ...state,
        action.item,
      ];
    }
    case UPDATE_ITEM_POUCH: {
      const indexOfId = state.findIndex(item => item._id === action.item._id);
      if (indexOfId === -1) return state;
      return [
        ...state.slice(0, indexOfId),
        action.item,
        ...state.slice(indexOfId + 1),
      ];
    }
    case CREATE_ITEM: {
      const newState = ensureRootAndInboxExists(state, action);
      const parentProjectIndex = newState.findIndex(item => item._id === action.parentProjectId);
      return [
        ...newState.slice(0, parentProjectIndex),
        { ...newState[parentProjectIndex],
          children: [
            ...newState[parentProjectIndex].children,
            action.item._id,
          ] },
        ...newState.slice(parentProjectIndex + 1),
        action.item,
      ];
    }
    case COMPLETE_TASK: {
      const indexOfId = state.findIndex(item => item._id === action.id);
      if (indexOfId === -1) return state;
      return [
        ...state.slice(0, indexOfId),
        {
          ...state[indexOfId],
          isCompleted: action.isCompleted,
        },
        ...state.slice(indexOfId + 1),
      ];
    }
    case DELETE_TASK: {
      const parentProject = state.find(item => item._id === action.parentProjectId);
      const childIndexInParent = parentProject.children.indexOf(action.id);
      if (childIndexInParent === -1) return state;
      const taskIndex = state.findIndex(item => item._id === action.id);
      const stateWithoutTask = [
        ...state.slice(0, taskIndex),
        ...state.slice(taskIndex + 1),
      ];
      const parentProjectIndex = stateWithoutTask.findIndex(item =>
        item._id === action.parentProjectId);
      return [
        ...stateWithoutTask.slice(0, parentProjectIndex),
        { ...parentProject,
          children: [
            ...parentProject.children.slice(0, childIndexInParent),
            ...parentProject.children.slice(childIndexInParent + 1),
          ] },
        ...stateWithoutTask.slice(parentProjectIndex + 1),
      ];
    }
    case DELETE_PROJECT: {
      const parentProject = state.find(item => item._id === action.parentProjectId);
      const childIndexInParent = parentProject.children.indexOf(action.id);
      if (childIndexInParent === -1) return state;
      const childIndex = state.findIndex(item => item._id === action.id);
      if (state[childIndex].children.length > 0) return state;
      const stateWithoutTask = [
        ...state.slice(0, childIndex),
        ...state.slice(childIndex + 1),
      ];
      const parentProjectIndex = stateWithoutTask.findIndex(item =>
        item._id === action.parentProjectId);
      return [
        ...stateWithoutTask.slice(0, parentProjectIndex),
        { ...parentProject,
          children: [
            ...parentProject.children.slice(0, childIndexInParent),
            ...parentProject.children.slice(childIndexInParent + 1),
          ] },
        ...stateWithoutTask.slice(parentProjectIndex + 1),
      ];
    }
    case UPDATE_ITEM: {
      const indexOfOldItem = state.findIndex(item => item._id === action.item._id);
      const oldItem = state[indexOfOldItem];
      let newItem;
      if (oldItem.isProject) {
        newItem = {
          ...oldItem,
          name: action.item.name,
        };
      } else {
        newItem = {
          ...oldItem,
          name: action.item.name,
          tags: action.item.tags,
        };
      }
      return [
        ...state.slice(0, indexOfOldItem),
        newItem,
        ...state.slice(indexOfOldItem + 1),
      ];
    }
    case MOVE_ITEM: {
      const itemIndex = state.findIndex(item => item._id === action.id);
      const previousParentIndex = state.findIndex(item => item._id === action.previousParent);
      const newParentIndex = state.findIndex(item => item._id === action.newParent);
      if (itemIndex === -1 || previousParentIndex === -1 || newParentIndex === -1) return state;
      if (!state[newParentIndex].isProject || !state[previousParentIndex].isProject) return state;
      const previousParent = state[previousParentIndex];
      const childIndexInPreviousParent = previousParent.children.indexOf(action.id);
      if (childIndexInPreviousParent === -1) return state;
      const stateWithoutItemInPreviousParent = [
        ...state.slice(0, previousParentIndex),
        { ...previousParent,
          children: [
            ...previousParent.children.slice(0, childIndexInPreviousParent),
            ...previousParent.children.slice(childIndexInPreviousParent + 1),
          ] },
        ...state.slice(previousParentIndex + 1),
      ];
      const newParent = state[newParentIndex];
      return [
        ...stateWithoutItemInPreviousParent.slice(0, newParentIndex),
        { ...newParent, children: [...newParent.children, action.id] },
        ...stateWithoutItemInPreviousParent.slice(newParentIndex + 1),
      ];
    }
    default: {
      return state;
    }
  }
}
