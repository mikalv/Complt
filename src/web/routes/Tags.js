import React from 'react';
import { connect } from 'react-redux';
import List from 'preact-material-components/List';
import LinkListItem from '../components/LinkListItem';
import getFilteredItems from '../../common/utils/getFilteredItems';
import areInitialItemsLoaded from '../../common/utils/areInitialItemsLoaded';
import Loading from '../components/Loading';

export const Tags = props => (
  <List className="flex-child">
    {props.tags.map(tag =>
      <LinkListItem
        to={`/tag/${tag}`}
        key={tag}
      >{tag}</LinkListItem>,
    )}
  </List>
);

export function mapStateToProps(state) {
  const tags = [];
  getFilteredItems(state.items, state.itemsToShow).forEach((item) => {
    if (!Array.isArray(item.tags)) return;
    item.tags.forEach((tag) => {
      if (tags.indexOf(tag) === -1) tags.push(tag);
    });
  });
  return { tags };
}

export default areInitialItemsLoaded(connect(mapStateToProps)(Tags), Loading);
