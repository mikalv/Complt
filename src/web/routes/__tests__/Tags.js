import React from 'react';
import { mount } from 'enzyme';
import { Tags, mapStateToProps } from '../Tags';
import items from '../../exampleItems';

describe('Tags component', () => {
  it('renders without crashing', () => {
    mount(<Tags tags={[]} />);
  });
  it('renders some items correctly', () => {
    const component = mount(<Tags tags={mapStateToProps({ items }).tags} />);
    expect(component).toMatchSnapshot();
  });
});
