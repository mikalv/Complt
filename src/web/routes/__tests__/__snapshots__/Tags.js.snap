// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags component renders some items correctly 1`] = `
<Component
  tags={
    Array [
      "@tag",
      "@tag2",
      "@tag3",
    ]
  }
>
  <List>
    <ul
      className="md-list"
    >
      <ListItem
        activeClassName="md-text--theme-primary"
        component={[Function]}
        expanderIconChildren="keyboard_arrow_down"
        primaryText="@tag"
        to="/tag/@tag"
      >
        <li
          className="md-list-item"
        >
          <AccessibleFakeButtonInked
            __SUPER_SECRET_REF__={[Function]}
            aria-expanded={null}
            className="md-list-tile md-text"
            component={[Function]}
            inkTransitionEnterTimeout={450}
            inkTransitionLeaveTimeout={300}
            inkTransitionOverlap={150}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            to="/tag/@tag"
          >
            <AccessibleFakeButton
              aria-expanded={null}
              className="md-list-tile md-text"
              component={[Function]}
              ink={
                <InkContainer
                  className={undefined}
                  disabledInteractions={undefined}
                  inkClassName={undefined}
                  inkStyle={undefined}
                  style={undefined}
                  transitionEnterTimeout={450}
                  transitionLeaveTimeout={300}
                  transitionOverlap={150}
                  waitForInkTransition={undefined}
                />
              }
              onClick={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseLeave={[Function]}
              onMouseOver={[Function]}
              onTouchEnd={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex={0}
              to="/tag/@tag"
            >
              <Link
                aria-expanded={null}
                aria-pressed={false}
                className="md-fake-btn md-pointer--hover md-list-tile md-text"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                onlyActiveOnIndex={false}
                role="button"
                style={Object {}}
                tabIndex={0}
                to="/tag/@tag"
              >
                <a
                  aria-expanded={null}
                  aria-pressed={false}
                  className="md-fake-btn md-pointer--hover md-list-tile md-text"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onTouchEnd={[Function]}
                  onTouchStart={[Function]}
                  role="button"
                  style={Object {}}
                  tabIndex={0}
                >
                  <InkContainer
                    transitionEnterTimeout={450}
                    transitionLeaveTimeout={300}
                    transitionOverlap={150}
                  >
                    <ReactTransitionGroup
                      childFactory={[Function]}
                      className="md-ink-container"
                      component="div"
                    >
                      <div
                        className="md-ink-container"
                      />
                    </ReactTransitionGroup>
                  </InkContainer>
                  <TileAddon
                    activeClassName="md-text--theme-primary"
                  />
                  <ListItemText
                    activeClassName="md-text--theme-primary"
                    className=""
                    primaryText="@tag"
                  >
                    <div
                      className="md-tile-content"
                    >
                      <div
                        className="md-tile-text--primary md-text"
                      >
                        @tag
                      </div>
                    </div>
                  </ListItemText>
                  <TileAddon
                    activeClassName="md-text--theme-primary"
                  />
                </a>
              </Link>
            </AccessibleFakeButton>
          </AccessibleFakeButtonInked>
        </li>
      </ListItem>
      <ListItem
        activeClassName="md-text--theme-primary"
        component={[Function]}
        expanderIconChildren="keyboard_arrow_down"
        primaryText="@tag2"
        to="/tag/@tag2"
      >
        <li
          className="md-list-item"
        >
          <AccessibleFakeButtonInked
            __SUPER_SECRET_REF__={[Function]}
            aria-expanded={null}
            className="md-list-tile md-text"
            component={[Function]}
            inkTransitionEnterTimeout={450}
            inkTransitionLeaveTimeout={300}
            inkTransitionOverlap={150}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            to="/tag/@tag2"
          >
            <AccessibleFakeButton
              aria-expanded={null}
              className="md-list-tile md-text"
              component={[Function]}
              ink={
                <InkContainer
                  className={undefined}
                  disabledInteractions={undefined}
                  inkClassName={undefined}
                  inkStyle={undefined}
                  style={undefined}
                  transitionEnterTimeout={450}
                  transitionLeaveTimeout={300}
                  transitionOverlap={150}
                  waitForInkTransition={undefined}
                />
              }
              onClick={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseLeave={[Function]}
              onMouseOver={[Function]}
              onTouchEnd={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex={0}
              to="/tag/@tag2"
            >
              <Link
                aria-expanded={null}
                aria-pressed={false}
                className="md-fake-btn md-pointer--hover md-list-tile md-text"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                onlyActiveOnIndex={false}
                role="button"
                style={Object {}}
                tabIndex={0}
                to="/tag/@tag2"
              >
                <a
                  aria-expanded={null}
                  aria-pressed={false}
                  className="md-fake-btn md-pointer--hover md-list-tile md-text"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onTouchEnd={[Function]}
                  onTouchStart={[Function]}
                  role="button"
                  style={Object {}}
                  tabIndex={0}
                >
                  <InkContainer
                    transitionEnterTimeout={450}
                    transitionLeaveTimeout={300}
                    transitionOverlap={150}
                  >
                    <ReactTransitionGroup
                      childFactory={[Function]}
                      className="md-ink-container"
                      component="div"
                    >
                      <div
                        className="md-ink-container"
                      />
                    </ReactTransitionGroup>
                  </InkContainer>
                  <TileAddon
                    activeClassName="md-text--theme-primary"
                  />
                  <ListItemText
                    activeClassName="md-text--theme-primary"
                    className=""
                    primaryText="@tag2"
                  >
                    <div
                      className="md-tile-content"
                    >
                      <div
                        className="md-tile-text--primary md-text"
                      >
                        @tag2
                      </div>
                    </div>
                  </ListItemText>
                  <TileAddon
                    activeClassName="md-text--theme-primary"
                  />
                </a>
              </Link>
            </AccessibleFakeButton>
          </AccessibleFakeButtonInked>
        </li>
      </ListItem>
      <ListItem
        activeClassName="md-text--theme-primary"
        component={[Function]}
        expanderIconChildren="keyboard_arrow_down"
        primaryText="@tag3"
        to="/tag/@tag3"
      >
        <li
          className="md-list-item"
        >
          <AccessibleFakeButtonInked
            __SUPER_SECRET_REF__={[Function]}
            aria-expanded={null}
            className="md-list-tile md-text"
            component={[Function]}
            inkTransitionEnterTimeout={450}
            inkTransitionLeaveTimeout={300}
            inkTransitionOverlap={150}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            to="/tag/@tag3"
          >
            <AccessibleFakeButton
              aria-expanded={null}
              className="md-list-tile md-text"
              component={[Function]}
              ink={
                <InkContainer
                  className={undefined}
                  disabledInteractions={undefined}
                  inkClassName={undefined}
                  inkStyle={undefined}
                  style={undefined}
                  transitionEnterTimeout={450}
                  transitionLeaveTimeout={300}
                  transitionOverlap={150}
                  waitForInkTransition={undefined}
                />
              }
              onClick={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseLeave={[Function]}
              onMouseOver={[Function]}
              onTouchEnd={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex={0}
              to="/tag/@tag3"
            >
              <Link
                aria-expanded={null}
                aria-pressed={false}
                className="md-fake-btn md-pointer--hover md-list-tile md-text"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                onlyActiveOnIndex={false}
                role="button"
                style={Object {}}
                tabIndex={0}
                to="/tag/@tag3"
              >
                <a
                  aria-expanded={null}
                  aria-pressed={false}
                  className="md-fake-btn md-pointer--hover md-list-tile md-text"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onTouchEnd={[Function]}
                  onTouchStart={[Function]}
                  role="button"
                  style={Object {}}
                  tabIndex={0}
                >
                  <InkContainer
                    transitionEnterTimeout={450}
                    transitionLeaveTimeout={300}
                    transitionOverlap={150}
                  >
                    <ReactTransitionGroup
                      childFactory={[Function]}
                      className="md-ink-container"
                      component="div"
                    >
                      <div
                        className="md-ink-container"
                      />
                    </ReactTransitionGroup>
                  </InkContainer>
                  <TileAddon
                    activeClassName="md-text--theme-primary"
                  />
                  <ListItemText
                    activeClassName="md-text--theme-primary"
                    className=""
                    primaryText="@tag3"
                  >
                    <div
                      className="md-tile-content"
                    >
                      <div
                        className="md-tile-text--primary md-text"
                      >
                        @tag3
                      </div>
                    </div>
                  </ListItemText>
                  <TileAddon
                    activeClassName="md-text--theme-primary"
                  />
                </a>
              </Link>
            </AccessibleFakeButton>
          </AccessibleFakeButtonInked>
        </li>
      </ListItem>
    </ul>
  </List>
</Component>
`;
